### DISCUSSION TITLE
source_nodes is blank when using FunctionCallingAgentWorker in llama-index

### DISCUSSION THREAD

**Discussion by jkglangchain** [community/community#135754] [Programming Help]
### Select Topic Area

Question

### Body

`
from llama_index.core.agent import FunctionCallingAgentWorker
from llama_index.core.agent import AgentRunner

agent_worker = FunctionCallingAgentWorker.from_tools(
    tool_retriever=obj_retriever,
    llm=llm, 
    system_prompt=""" \
You are an agent designed to answer queries over a set of given papers.
Please always use the tools provided to answer a question. Do not rely on prior knowledge.\

""",
    verbose=True
)
agent = AgentRunner(agent_worker)

response = agent.query(
    "ask questions"
)

print(response)`

Response(response='desired answer', source_nodes=[], metadata=None)

when i use this function , i want the sources_nodes to fetch the metadata from which it referred

**Comment by ebndev** [comment]
Hi @jkglangchain, thanks for posting in GitHub Discussions! 

I've gone ahead and moved this post to our `Programming Help` category, as this topic may be more relevant there and might help someone give you a nudge in the right direction. 😁

**Comment by erencanbas** [comment]
Adjust AgentWorker to collect source nodes. The **FunctionCallingAgentWorker**  should be configured to collect and return **source_nodes** with metadata.
```
from llama_index.core.agent import FunctionCallingAgentWorker
from llama_index.core.agent import AgentRunner


agent_worker = FunctionCallingAgentWorker.from_tools(
    tool_retriever=obj_retriever,
    llm=llm,
    system_prompt="""
    You are an agent designed to answer queries over a set of given papers.
    Please always use the tools provided to answer a question. Do not rely on prior knowledge.
    """,
    verbose=True
)
agent = AgentRunner(agent_worker)

response = agent.query("ask questions")

print("Response:", response.response)
print("Source Nodes:", response.source_nodes)
print("Metadata:", response.metadata)

for node in response.source_nodes:
    print("Source Node Metadata:", node.metadata)
```
`

**Reply by jkglangchain** [reply]
Response(response='answer', source_nodes=[], metadata=None)

Assume above is the response i got , so value are blanks . Printing it separately doesn't help. Is it to do with the llm which I'm using ? Im using Azureopenai instead of openai.


from llama_index.llms.azure_openai import AzureOpenAI

llm = AzureOpenAI(
    model="",
    deployment_name="",
    api_key="",
    azure_endpoint="",
    api_version=""
)

**Reply by erencanbas** [reply]
**Check LLM API Ca**ll:

```
response = llm.query("your prompt")
print(response)
```

**Verbose Debugging:**
Since you've set verbose=True, check the logs or outputs during the agent’s execution. If there are any errors or issues with connecting to Azure OpenAI, they should be highlighted there.

If possible, test the setup using a different LLM to see if the issue is specific to Azure OpenAI or more generally related to the configuration.

Ensure that your **obj_retriever** is working correctly and returning the data needed for the agent to process. If the retriever isn’t working as expected, the agent might not have any data to answer from, resulting in blank responses.
